PLATFORM-4222 | upgrade to PHP 7.3.7 on dev

macbre opened the pull request:
https://wikia-inc.atlassian.net/browse/PLATFORM-4222
https://wikia-inc.atlassian.net/browse/CORE-260

https://www.php.net/releases/7_3_7.php

```
$ docker run -it --rm artifactory.wikia-inc.com/platform/php-wikia-devbox:latest -v
PHP 7.3.7 (fpm-fcgi) (built: Jul 17 2019 22:34:16)
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.7, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.7, Copyright (c) 1999-2018, by Zend Technologies
    with Xdebug v2.7.2, Copyright (c) 2002-2019, by Derick Rethans
```

### This pull request updates

* PHP to v7.3.7
* PHP extensions in base image: mustache to v0.8.1 and tideways to v4.1.7
* PHP extensions for dev environment: uopz to v6.01, xdebug to v2.7.2 and yaml to v2.0.4

### Additional changes

* PHP 7.3 has fixed handling of logging to stdout from fpm workers. We got rid of a #15173 hack involving a small container called `mediawiki-logger` + MediaWiki logging via socket

### Performance

PHP 7.3 is ~12% faster than 7.0 (median response time for ~1.5k various URLs).

macbre committed:
CORE-260 | Base PHP 7.3.3 image builds fine

mustache and tideways extensions are now in never versions

macbre committed:
CORE-260 | update dev PHP extensions for PHP 7.3.3

macbre committed:
Travis | run tests using PHP 7.3

macbre committed:
dev/docker-compose.yml: hardcode DB nodes IP addresses

macbre committed:
OutputPage::makeResourceLoaderLink | cast to array to avoid "Warning: count(): Parameter must be an array or an object that implements Countable"

macbre committed:
wgSkinExtensionFunctions: is gone, no default in config, no variable in WikiFactory

macbre committed:
includes/json/FormatJson.php: backport the latest version from MediaWiki upstream

macbre committed:
CategoryHelper: use break instead of continue

macbre committed:
CategoryHelper: fix "Parameter must be an array or an object that implements Countable"

macbre committed:
Merge branch 'dev' into CORE-260-updgrade-to-php-7.3.3

legoktm committed:
LocalisationCache: Avoid use of compact()

In PHP 7.3, compact() now raises notices if the variable is undefined, which
is something that we expect. So we can check whether the key exists instead
of bothering with compat() and suppressing warnings.

Backporting https://github.com/wikimedia/mediawiki/commit/d0463178dfa09b79b3a08fee939da1beed030824 from MediaWiki 1.32

macbre committed:
Updating wikia/simplehtmldom (v1.6.2 => v1.8.1)

Try to fix

PHP Warning:  preg_match(): Compilation failed: invalid range in character class at offset 4 in app/lib/composer/wikia/simplehtmldom/simple_html_dom.php on line 1385

macbre committed:
Travis: fetch mustache v0.8.1 tgz file and install it from a file

Failed to download pecl/mustache, version "0.8.1", latest release is version 0.8.0, stability "stable", use "channel://pecl.php.net/mustache-0.8.0" to install
install failed

macbre committed:
Travis: pecl - force install of uopz

macbre committed:
Travis: try to install mustache at least

macbre committed:
Travis: try to install mustache at least

macbre committed:
Merge branch 'dev' into CORE-260-updgrade-to-php-7.3.3

macbre committed:
PHP 7.3 | mbereg_replace is deprecated, use mb_ereg_replace instead

https://wiki.php.net/rfc/deprecations_php_7_3

The following list provides a short overview of the functionality targeted for deprecation, while more detailed explanation is provided in the Proposal section:
* Undocumented mbstring function aliases

macbre committed:
Travis: try to install uopz from tgz file

macbre committed:
ArticleHTMLCleanupTest::testEmptyHeadersCleaner | we expect what we passed

macbre committed:
Travis | use Xenial dist

macbre committed:
Travis | remove uopz test skipping for PHP 7.2+

macbre committed:
PortableInfoboxMobileRenderServiceTest | do not emit DOCTYPE strings

macbre committed:
Merge branch 'dev' into CORE-260-updgrade-to-php-7.3.3

macbre committed:
Merge remote-tracking branch 'origin/dev' into CORE-260-updgrade-to-php-7.3.3

macbre committed:
Bump WIKIA_BASE_IMAGE_HASH

pwojcik86 commented on includes/OutputPage.php:
```diff
@@ -2745,6 +2745,8 @@ public function getResourceLoader() {
 	public function makeResourceLoaderLink( $modules, $only, $useESI = false, array $extraQuery = array(), $loadCall = false ) {
 		global $wgResourceLoaderUseESI;
 
+		$modules = (array) $modules; // Wikia change - PHP 7.3
```
Can `$modules` be declared as an `array` instead of this cast?

macbre replied:
I've simply "backported" the approach from MediaWiki's upstream - https://github.com/wikimedia/mediawiki/blob/c4e284f1130cdc2a4811120594bdb8013d17eaaa/includes/OutputPage.php#L3303

But sure, let's try adding a type hint in method declaration.

pwojcik86 replied:
I just read the comment for this method, and `$modules` is an `array` or `string`, so this is the only choice.

mszabo-wikia commented on .travis.yml:
```diff
@@ -1,24 +1,26 @@
-php: 7.0
-dist: trusty
+php: 7.3
+dist: xenial
```
👍

macbre committed:
wfSetupSession: suppress PHP warnings

These are reported for /api.php requests:

PHP Warning: session_cache_limiter(): Cannot change cache limiter when session is active in /usr/wikia/slot1/current/src/includes/GlobalFunctions.php on line 3479
PHP Warning: session_set_save_handler(): Cannot change save handler when session is active in /usr/wikia/slot1/current/src/includes/GlobalFunctions.php on line 3458
PHP Warning: session_set_cookie_params(): Cannot change session cookie parameters when session is active in /usr/wikia/slot1/current/src/includes/GlobalFunctions.php on line 3478

anomiex committed:
API: Account for PHP 7.2 change

PHP 7.2 broke existing functionality in making count( null ) raise a
warning. So add tests for null all over the place, or change tests where
we know the value is null or an array (but not false, empty-string, or
0) to just cast to boolean.

Backporting https://github.com/wikimedia/mediawiki/commit/9f8626100049a3cd37e23744ed235000c3fef3c8 from MediaWiki 1.32

Commit:
Fix XMPReader/XMPValidator for PHP 7

PHP Warning: count(): Parameter must be an array or an object that implements Countable in /usr/wikia/slot1/current/src/includes/media/XMP.php on line 701

Tripped up by old typos being caught in new warnings on
PHP 7.x: was checking a scalar's count when meant to
check its containing array's count, and was using a str
"\0" where a literal 0 should have been used (now considered
non-numeric).

Tested locally on 7.2

Backporting https://github.com/wikimedia/mediawiki/commit/f2f370cb31b0d2b9d88afa10161b39a1b87af093 from MediaWiki 1.32

macbre committed:
Follow | fix a fatal error, we tried to treat a string as an array

Unexpected non-MediaWiki exception encountered, of type "Error"
Error: [] operator not supported for strings in /usr/wikia/slot1/current/src/extensions/wikia/Follow/FollowHelper.class.php:34

macbre committed:
Merge remote-tracking branch 'origin/dev' into CORE-260-updgrade-to-php-7.3.3

macbre committed:
docker-compose | PHP 7.3 now correctly supports logging to stdout with prefixing and triming the JSON messages

macbre committed:
php-fpm | remove workarounds for broken fpm logging from PHP 7.0

Explicitily disable decorate_workers_output config variable and increase log limit to UDP packet size

macbre committed:
Logging | use LOG_STDOUT_ONLY instead of LOG_SOCKET_ONLY

macbre committed:
Logging | remove logging to socket hack for PHP 7.0

macbre committed:
Remove mediawiki-logger image

macbre committed:
LoggerFactoryTest | fix broken test

macbre committed:
Merge remote-tracking branch 'origin/dev' into CORE-260-updgrade-to-php-7.3.3

pwojcik86 committed:
Merge branch 'dev' into CORE-260-updgrade-to-php-7.3.3

themech committed:
Merge branch 'dev' into CORE-260-updgrade-to-php-7.3.3

themech committed:
Use php 7.3 on dev(boxes)

themech committed:
mediawiki-logger is no more

themech commented:
taking it over

themech committed:
Update base images versions

themech committed:
Bump xdebug to 2.7.2

themech committed:
bump yaml to 2.0.4

mszabo-wikia commented on docker/base/Dockerfile:
```diff
@@ -1,6 +1,6 @@
 # This is a base Docker image used by Wikia's MediaWiki app. It uses base PHP image provided by Docker with additional
 # production dependencies installed.
-FROM php:7.0.30-fpm-stretch
+FROM php:7.3.3-fpm-stretch
```
7.3.7 on buster is out

themech replied:
yeah, done in https://github.com/Wikia/app/pull/16873/commits/7f3999d4b05bfeccc44115b6d1bc1daaa9328de7

themech committed:
bump php to 7.3.7

themech committed:
remove lua experiment

themech committed:
we don't use phplint anymore

macbre commented:
:+1:

themech committed:
Bring back the soecket logger on prod

themech committed:
Merge pull request #17311 from Wikia/bring-back-the-socket-logger

Bring back the socket logger on prod

themech committed:
Don't use LOG_SOCKET_ADDRESS on dev

themech committed:
We don't need docker/dev anymore

themech committed:
travis: try using bionic

themech committed:
Revert "travis: try using bionic" - libmustache package is not ready yet

This reverts commit 2d2ab02ea71f002a7d71d1ac409cf099a114d4bc.

themech committed:
We don't use logger container on dev anymore

themech committed:
hotspots: count throws a warning when called with a null value

themech committed:
Fix the invalid range in character class warning (PHP 7.3)

mszabo-wikia reviewed (changes requested):
It seems we're missing some more backports (listed on https://phabricator.wikimedia.org/T206988):
* https://gerrit.wikimedia.org/r/c/mediawiki/core/+/45485 — warning in `LanguageConverter`
* https://gerrit.wikimedia.org/r/c/mediawiki/extensions/ParserFunctions/+/467306 — warning in `ExprParser`

themech committed:
PHP 7.3: port https://phabricator.wikimedia.org/T206974 from upstream

themech committed:
PHP 7.3" port https://gerrit.wikimedia.org/r/#/c/mediawiki/extensions/ParserFunctions/+/467306/

themech committed:
PHP 7.3: port https://gerrit.wikimedia.org/r/#/c/mediawiki/core/+/454852/ from upstream

themech committed:
PHP 7.3: port https://phabricator.wikimedia.org/T207100 from upstream

themech commented:
Good point @mszabo-wikia. I've ported the changes from upstream in the last 4 commits.
